### PR TITLE
Add a test to check for machineset leftovers ports

### PR DESCRIPTION
**Pull request by itzikb-redhat**
Create a machineset with a bogus serverGroupID
Delete the machineset
Check that there is no leftover port associated with the deleted machineset

### CODE REVIEW

**Comment by rlobillo** [comment]
Hello @itzikb-redhat I think you can move the test of servers.go as I think it is suitable there. For the file I created, I try to give "openstack" perspective (servers, servergroups, volumes...) and the operation you are testing is related to the creating of a new server. What do you think?

**Comment by itzikb-redhat** [comment]
> Hello @itzikb-redhat I think you can move the test to **servers.go** file. For the files I created until now, I tried to give "openstack" perspective (servers, servergroups, volumes...) and the operation you are testing is related to the creation of a new server. What do you think?

I think it's more related to machinesets but we can discuss

**Review by rlobillo — COMMENTED** [review]
Great job @itzikb-redhat 

I think it would be very useful in this complex testcase to make use of **g.By**. With this utility it is easier for maintainers to understand the different stages in the testcase. Would you mind to include them in the test? This will help for review too.

Also, in the **Expect** commands, it is possible to include a failure message: 

`o.Expect(err).NotTo(o.HaveOccurred(), "Error unmarshaling new MachineSet Specs")
`

I think that is very useful to get a clear view of the failure by only checking the logs.

[0] https://onsi.github.io/ginkgo/#documenting-complex-specs-by

File: test/extended/openstack/machinesets.go
Comment (rlobillo) [reply test/extended/openstack/machinesets.go:57]: I was referring to line 47 only. :)

File: test/extended/openstack/machinesets.go
Comment (MaysaMacedo) [inline test/extended/openstack/machinesets.go:46]: this line and the next one are unnecessary

File: test/extended/openstack/machinesets.go
Comment (MaysaMacedo) [inline test/extended/openstack/machinesets.go:68]: Is it needed to get the config again if it's already available at line 40 or it's a different one?

File: test/extended/openstack/machinesets.go
Comment (MaysaMacedo) [inline test/extended/openstack/machinesets.go:82]: what t represents? a machineset?

File: test/extended/openstack/machinesets.go
Comment (MaysaMacedo) [inline test/extended/openstack/machinesets.go:91]: it might be better to move this line to after 88 so we can easily identify which err is being checked

File: test/extended/openstack/machinesets.go
Comment (MaysaMacedo) [inline test/extended/openstack/machinesets.go:92]:
```suggestion
                        g.By("Create machineSet with new name and server group")
			ms, err := framework.CreateMachineSet(rclient, newMachinesetParams)
```

File: test/extended/openstack/machinesets.go
Comment (MaysaMacedo) [inline test/extended/openstack/machinesets.go:128]: is the name of the port exactly a fragment of the machineset name or it should be equal?
If it's equal, we could filter the ports query by that name as well

File: test/extended/openstack/machinesets.go
Comment (MaysaMacedo) [inline test/extended/openstack/machinesets.go:100]: What happens if 10 seconds was not enough to create the machine? Perhaps we should poll instead

**Review by MaysaMacedo — COMMENTED** [review]
Is this test an outcome of a bugzilla?

**Comment by itzikb-redhat** [comment]
> 

@MaysaMacedo [BZ2073398](https://bugzilla.redhat.com/show_bug.cgi?id=2073398)

File: test/extended/openstack/machinesets.go
Comment (itzikb-redhat) [reply test/extended/openstack/machinesets.go:46]: Done

File: test/extended/openstack/machinesets.go
Comment (itzikb-redhat) [reply test/extended/openstack/machinesets.go:57]: Done

**Comment by MaysaMacedo** [comment]
> > 
> 
> @MaysaMacedo [BZ2073398](https://bugzilla.redhat.com/show_bug.cgi?id=2073398)

@itzikb-redhat It would be good to add a comment to the code, linking that bz. Thanks

File: test/extended/openstack/machinesets.go
Comment (itzikb-redhat) [reply test/extended/openstack/machinesets.go:68]: I'm changing to e2e.LoadConfig , anyway cfg is not accessible here

File: test/extended/openstack/machinesets.go
Comment (itzikb-redhat) [reply test/extended/openstack/machinesets.go:82]: I modified the name to newProviderSpec, It's the providerspec of the machineset

File: test/extended/openstack/machinesets.go
Comment (itzikb-redhat) [reply test/extended/openstack/machinesets.go:91]: Done

File: test/extended/openstack/machinesets.go
Comment (itzikb-redhat) [reply test/extended/openstack/machinesets.go:92]: Add a similar one

File: test/extended/openstack/machinesets.go
Comment (itzikb-redhat) [reply test/extended/openstack/machinesets.go:128]: The ports contains the string but it's not exact

**Comment by itzikb-redhat** [comment]
> > > 
> > 
> > 
> > @MaysaMacedo [BZ2073398](https://bugzilla.redhat.com/show_bug.cgi?id=2073398)
> 
> @itzikb-redhat It would be good to add a comment to the code, linking that bz. Thanks

Done

**Comment by pierreprinetti** [comment]
Rebased on `main`

File: test/extended/openstack/machinesets.go
Comment (MaysaMacedo) [inline test/extended/openstack/machinesets.go:78]: this err also needs to be handled

File: test/extended/openstack/machinesets.go
Comment (MaysaMacedo) [inline test/extended/openstack/machinesets.go:109]: if an error happened once and eventually the machine was found, will we still have an err here and fail when it shouldn't?

File: test/extended/openstack/machinesets.go
Comment (MaysaMacedo) [inline test/extended/openstack/machinesets.go:113]: we probably need to poll for machine not found in this case

File: test/extended/openstack/machinesets.go
Comment (MaysaMacedo) [inline test/extended/openstack/machinesets.go:93]: is there a need for a defer if the machine is explicitly deleted in the test?

File: test/extended/openstack/machinesets.go
Comment (itzikb-redhat) [reply test/extended/openstack/machinesets.go:109]: If at the last iteration of the loop there was an error then it raises it , even there wasn't then it breaks out of the loop and the expect should be fine

File: test/extended/openstack/machinesets.go
Comment (itzikb-redhat) [reply test/extended/openstack/machinesets.go:93]: We need it in case the test errs before the machineset is deleted

File: test/extended/openstack/machinesets.go
Comment (itzikb-redhat) [reply test/extended/openstack/machinesets.go:78]: Done

File: test/extended/openstack/utils.go
Comment (dulek) [inline test/extended/openstack/utils.go:159]: Whoa, what a code duplication,. why? How about:

	if err != nil == shouldExist {
		retries += 1
		time.Sleep(time.Second * delay)
	} else {
		break
	}

Please double-check me if the logic is identical.

You should also check the error type here if it's NotFound.

File: test/extended/openstack/machinesets.go
Comment (dulek) [inline test/extended/openstack/machinesets.go:109]: This has to go before merge.

File: test/extended/openstack/machinesets.go
Comment (dulek) [inline test/extended/openstack/machinesets.go:85]: You should have a random name here instead.

**Review by dulek — CHANGES_REQUESTED** [review]
Looks pretty good, minor comments inline.

File: test/extended/openstack/utils.go
Comment (itzikb-redhat) [reply test/extended/openstack/utils.go:159]: agree! And logic is ok

File: test/extended/openstack/machinesets.go
Comment (itzikb-redhat) [reply test/extended/openstack/machinesets.go:109]: Done

File: test/extended/openstack/machinesets.go
Comment (itzikb-redhat) [reply test/extended/openstack/machinesets.go:85]: Done

File: test/extended/openstack/utils.go
Comment (itzikb-redhat) [reply test/extended/openstack/utils.go]: About the Notfound - I'm checking it in the test itself, do you think I should add the logic in the function?

Comment (dulek) [reply test/extended/openstack/utils.go]: Nah, okay, I see NotFound check now.

**Comment by itzikb-redhat** [comment]
> Looks good. One more thing is that it's a good practice to put vendoring changes into a separate commit in the PR, so that later on we can revert the test without necessarily reverting the `go mod` update. Please update that.

Done. Please let me know if it seems ok

**Comment by dulek** [comment]
@itzikb-redhat: Please do not retest this, looks like we've found an interference between two tests:

```
{  fail [github.com/openshift/openstack-test/test/extended/openstack/servers.go:114]: Error gathering Openstack info for machine bogus-2881xjhrl-mbq69
Unexpected error:
    <gophercloud.ErrDefault404>: {
        ErrUnexpectedResponseCode: {
            BaseError: {DefaultErrString: "", Info: ""},
            URL: "https://compute.rdo.mtl2.vexxhost.net/v2.1/0693e2bb538c42b79a49fe6d2e61b0fc/servers/",
            Method: "GET",
            Expected: [200, 203],
            Actual: 404,
            Body: "{\"message\": \"The resource could not be found.<br /><br />\\n\\n\\n\", \"code\": \"404 Not Found\", \"title\": \"Not Found\"}",
        },
    }
    Resource not found
occurred

I guess [here](https://github.com/openshift/openstack-test/blob/61409a88ec92dddabb836be062cbd2ed77a248ab/test/extended/openstack/servers.go#L104-L118) either a new Machine should be created, or only installer-created worker nodes should be used. Or we should just catch 404 and ignore it, later on checking if we've found at least one Machine. Please note that filtering to only have Machines in "Provisioned" status will not work, the state of the machine can easily change between fetching list and calling OpenStack.

**Comment by itzikb-redhat** [comment]
@dulek Thanks, @rlobillo saw that issue. We'll propose a patch.

**Comment by MaysaMacedo** [comment]
/hold
CCM was not deployed, likely depends on https://github.com/openshift/cluster-cloud-controller-manager-operator/pull/202

**Comment by MaysaMacedo** [comment]
/hold cancel
dependent PR was merged

**Comment by MaysaMacedo** [comment]
/approve remove
seems other test is being impacted

**Comment by itzikb-redhat** [comment]
So "[sig-installer][Suite:openshift/openstack] The OpenStack platform on instance creation should include the addresses on the machine specs [Suite:openshift/conformance/parallel]"
 failed

The updated PS should solve it for now but I wonder if the test shouldn't be modified to reflect the bug , i.e. to add a FIP to a machine and then check the addresses of the machine instead of going over all the machines

**Pull request merged**
